### ISSUE TITLE
mscoco pretrained weights uncorrect！

### ISSUE THREAD

**Issue by luckycallor** [xuannianz/EfficientDet#146]
I downloaded the mscoco pretrained weights (d0, d1 and d2), and test  with the script inference.py. The output scores are all very small, around 0.01.  I found that the model can output different scores with the same input and model weights. I guess there is something wrong in the pretrained weights, and the script inference.py can not load all model weights correctly. Can you check this please?

**Comment by SiBensberg** [comment]
did you set weighted Bifpn=True in inference.py?

**Comment by SiBensberg** [comment]
I tried D0 and it works:
![horse](https://user-images.githubusercontent.com/55738346/79040197-1f3d6300-7be7-11ea-848d-75c223362b7e.png)

with this in inference
```
def main():
    os.environ['CUDA_VISIBLE_DEVICES'] = '0'

    phi = 0
    weighted_bifpn = True
    model_path = 'test/efficientdet-d0.h5'
    image_sizes = (512, 640, 768, 896, 1024, 1280, 1408)
    image_size = image_sizes[phi]
    # coco classes
    classes = {value['id'] - 1: value['name'] for value in json.load(open('coco_90.json', 'r')).values()}
    num_classes = 90
    score_threshold = 0.3

```

**Comment by murdockhou** [comment]
@SiBensberg Could you help me check out my code? I want to use efficientdet-d0 to predict on single image but get nothing, and the scores of the mode predict are all **-1**. Here is my code:
```
from model import efficientdet
import cv2
import os 
import numpy as np 
import time 
import argparse
from utils.anchors import anchors_for_shape

num_classes = 90
score_threshold = 0.1
colors = [np.random.randint(0, 256, 3).tolist() for i in range(num_classes)]

def infer(model, img, anchors):
    src_image = img.copy()
    h, w = src_image.shape[:2]
    image, scale, offset_h, offset_w = preprocess(img, 512)

    start = time.time()
    boxes, scores, labels = model.predict_on_batch(
        [np.expand_dims(image, axis=0), np.expand_dims(anchors, axis=0)]
    )
    end = time.time()
    print ('infer time == ', end-start)

    boxes[0, :, [0, 2]] = boxes[0, :, [0, 2]] - offset_w
    boxes[0, :, [1, 3]] = boxes[0, :, [1, 3]] - offset_h
    boxes /= scale
    boxes[0, :, 0] = np.clip(boxes[0, :, 0], 0, w - 1)
    boxes[0, :, 1] = np.clip(boxes[0, :, 1], 0, h - 1)
    boxes[0, :, 2] = np.clip(boxes[0, :, 2], 0, w - 1)
    boxes[0, :, 3] = np.clip(boxes[0, :, 3], 0, h - 1)

    print (scores)

    # select indices which have a score above the threshold
    indices = np.where(scores[0, :] > score_threshold)[0]

    # select those detections
    boxes = boxes[0, indices]
    scores = scores[0, indices]
    labels = labels[0, indices]

    for box, score, label in zip(boxes, scores, labels):
        xmin = int(round(box[0]))
        ymin = int(round(box[1]))
        xmax = int(round(box[2]))
        ymax = int(round(box[3]))
        score = '{:.4f}'.format(score)
        class_id = int(label)
        color = colors[class_id]
        # class_name = classes[class_id]
        class_name = class_id
        label = '-'.join([class_name, score])
        ret, baseline = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.5, 1)
        cv2.rectangle(src_image, (xmin, ymin), (xmax, ymax), color, 1)
        cv2.rectangle(src_image, (xmin, ymax - ret[1] - baseline), (xmin + ret[0], ymax), color, -1)
        cv2.putText(src_image, label, (xmin, ymax - baseline), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 0), 1)
    
    return src_image

 

def preprocess(image, image_size):
    image_height, image_width = image.shape[:2]
    if image_height > image_width:
        scale = image_size / image_height
        resized_height = image_size
        resized_width = int(image_width * scale)
    else:
        scale = image_size / image_width
        resized_height = int(image_height * scale)
        resized_width = image_size
    image = cv2.resize(image, (resized_width, resized_height))
    new_image = np.ones((image_size, image_size, 3), dtype=np.float32) * 128.
    offset_h = (image_size - resized_height) // 2
    offset_w = (image_size - resized_width) // 2
    new_image[offset_h:offset_h + resized_height, offset_w:offset_w + resized_width] = image.astype(np.float32)
    new_image /= 255.
    mean = [0.485, 0.456, 0.406]
    std = [0.229, 0.224, 0.225]
    new_image[..., 0] -= mean[0]
    new_image[..., 1] -= mean[1]
    new_image[..., 2] -= mean[2]
    new_image[..., 0] /= std[0]
    new_image[..., 1] /= std[1]
    new_image[..., 2] /= std[2]
    return new_image, scale, offset_h, offset_w

if __name__ == '__main__':

    os.environ['CUDA_VISIBLE_DEVICES'] = '0'
    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '1'

    parser = argparse.ArgumentParser()
    parser.add_argument('--img', type=str, default='/media/hsw/E/datasets/crowdpose/crowdpose_images')
    parser.add_argument('--model', type=str, default='weights/efficientdet-d0.h5')
    args = parser.parse_args()

    model, prediction_model = efficientdet(phi=0, num_classes=num_classes, weighted_bifpn=True, score_threshold=score_threshold)

    prediction_model.load_weights(args.model, by_name=True)
    print ("Successfully load pretrained model")

    anchors = anchors_for_shape((512,512))

    for img_name in os.listdir(args.img):
        img = cv2.imread(os.path.join(args.img, img_name))
        img = infer(prediction_model, img, anchors)
        cv2.imshow('result', img)
        key = cv2.waitKey(0)
        if key == ord('q'):
            break

```

Is there some wrong? I used coco pretrained model. 
Env:

- tensorflow 1.15.1
- cuda 10.1
- python 3.6

Thank you very much.

**Comment by luckycallor** [comment]
> I tried D0 and it works:
> ![horse](https://user-images.githubusercontent.com/55738346/79040197-1f3d6300-7be7-11ea-848d-75c223362b7e.png)
> 
> with this in inference
> 
> ```
> def main():
>     os.environ['CUDA_VISIBLE_DEVICES'] = '0'
> 
>     phi = 0
>     weighted_bifpn = True
>     model_path = 'test/efficientdet-d0.h5'
>     image_sizes = (512, 640, 768, 896, 1024, 1280, 1408)
>     image_size = image_sizes[phi]
>     # coco classes
>     classes = {value['id'] - 1: value['name'] for value in json.load(open('coco_90.json', 'r')).values()}
>     num_classes = 90
>     score_threshold = 0.3
> ```

what is your ternsorflow and keras version?

**Comment by luckycallor** [comment]
@murdockhou I have the same problem with yours.

**Comment by murdockhou** [comment]
@luckycallor Have you solved this problem? I do not want use the `inference.py` provided by author, because I want to predict on mine image or video. But get nothing and the scores predicted are all **-1**. The complete code is above, alos tensorflow version(keras is embeded within tensorflow).

**Comment by luckycallor** [comment]
@murdockhou I used the inference.py provided by author, but have the same problem with yours. I have not solved it yet.

**Comment by luckycallor** [comment]
@SiBensberg Yes, I setted the weighted Bifpn=True, but get nothing, and the scores of the mode predict are all -1. Can you tell me more details about you code and run envs. Here is my code:

`import cv2
import json
import numpy as np
import os
import time
import glob

from model import efficientdet
from utils import preprocess_image, postprocess_boxes
from utils.draw_boxes import draw_boxes

def main():
    # os.environ['CUDA_VISIBLE_DEVICES'] = '0'

    phi = 0
    weighted_bifpn = True
    model_path = 'pretrained/efficientdet-d0.h5'
    image_path = 'test_image/tDQGinHoJ0_small.jpg'
    image_sizes = (512, 640, 768, 896, 1024, 1280, 1408)
    image_size = image_sizes[phi]
    # coco classes
    classes = {value['id'] - 1: value['name'] for value in json.load(open('coco_90.json', 'r')).values()}
    num_classes = 90
    score_threshold = 0.3
    colors = [np.random.randint(0, 256, 3).tolist() for _ in range(num_classes)]
    _, model = efficientdet(phi=phi,
                            weighted_bifpn=weighted_bifpn,
                            num_classes=num_classes,
                            score_threshold=score_threshold)
    model.load_weights(model_path, by_name=True)

    # for image_path in glob.glob('datasets/VOC2007/JPEGImages/*.jpg'):
    image = cv2.imread(image_path)
    src_image = image.copy()
    # BGR -> RGB
    image = image[:, :, ::-1]
    h, w = image.shape[:2]

    image, scale = preprocess_image(image, image_size=image_size)
    # run network
    start = time.time()
    boxes, scores, labels = model.predict_on_batch([np.expand_dims(image, axis=0)])
    boxes, scores, labels = np.squeeze(boxes), np.squeeze(scores), np.squeeze(labels)
    print(time.time() - start)
    boxes = postprocess_boxes(boxes=boxes, scale=scale, height=h, width=w)

    # select indices which have a score above the threshold
    indices = np.where(scores[:] > score_threshold)[0]

    # select those detections
    boxes = boxes[indices]
    labels = labels[indices]

    draw_boxes(src_image, boxes, scores, labels, colors, classes)

    cv2.imwrite('test_image/res.jpg', src_image)

if __name__ == '__main__':
    main()`

The env is:
python3.6.4
tensorflow 1.14.0
keras 2.2.4

**Comment by SiBensberg** [comment]
I actually changed nothing else in the code. I made a fresh conda environment with 

- tensorflow-gpu 1.15
- cuda 10.0.130
- cudnn 7.6.5

However I found 2 similiar issues to yours, you could try the solutions they provided. I hope that helps.
#126 
#108
